Enable landlock in cloud-hypervisor

praveen-pk opened the pull request:
I am sending this PR for some early feedback on my current approach for enabling Landlock in Cloud-Hypervisor. 

# Approach

I explicitly enable landlock in `event-monitor`, `http-server`, `signal-handler` as these threads are spawned before `vmm` thread. By the time these 3 threads are spawned, all the necessary FDs these threads need have already been opened. So, I just create landlock object and invoke `restrict_self` to sandbox these threads.

In the `vmm` thread, I enable landlock in `vm_create` method. Doing this will allow us to support both the invocation modes:

1) All the vm params are passed directly at cmdline
2) guest is started with a socket interface and vm_create request is sent at a later time.
 
All the paths from VMConfig and LandlockConfig are appended to ruleset before the `vmm` thread is sandboxed.  Rest of the threads (`serial-manager`, `vcpu*`, `_disk*`, `_net*`, `__rng`) are all spawned by `vmm` thread. So, they will automatically inherit the ruleset applied in `vmm` thread.


# Known TODOs:
* Enable Landlock in `vm_restore`
* Handle other configs that take file paths as listed in https://github.com/cloud-hypervisor/cloud-hypervisor/issues/5170#issuecomment-1917656408.  **[DONE]**
* Unit Tests
* Integration tests
* Address Linting errors

# Current Testing
## Step1
```
./target/debug/cloud-hypervisor \
    --api-socket /tmp/ch-socket \
    --kernel ${FW_IMAGE} \
        --disk path=${DISK_IMAGE1}  path=${DISK_IMAGE2} \
        --cpus boot=2 \
        --memory size=1024M \
        --serial tty \
        --console off \
        --landlock true
```
To this VM if a new disk is hot-added with:
```
 curl --unix-socket  /tmp/ch-socket  -i \
     -X PUT 'http://localhost/api/v1/vm.add-disk'  \
     -H 'Accept: application/json'               \
     -H 'Content-Type: application/json'         \
     -d '{
         "path":"/tmp/test.img"
         }'
HTTP/1.1 500
Server: Cloud Hypervisor API
Connection: keep-alive
Content-Type: application/json
Content-Length: 102

Error from device manager: Disk(Os { code: 13, kind: PermissionDenied, message: "Permission denied" })
```
## Step2

```
./target/debug/cloud-hypervisor \
    --api-socket /tmp/ch-socket \
    --kernel ${FW_IMAGE} \
        --disk path=${DISK_IMAGE1}  path=${DISK_IMAGE2} \
        --cpus boot=2 \
        --memory size=1024M \
        --serial tty \
        --console off \
        --landlock true \
        --landlock-rules path=/tmp/test.img,flags=rw
```
By passing the additional `landlock-rules` argument, hot-add request now passes:

```
$ curl --unix-socket  /tmp/ch-socket  -i      -X PUT 'http://localhost/api/v1/vm.add-disk'       -H 'Accept: application/json'
  -H 'Content-Type: application/json'              -d '{
         "path":"/tmp/test.img"
         }'
HTTP/1.1 200
Server: Cloud Hypervisor API
Connection: keep-alive
Content-Type: application/json
Content-Length: 36

{"id":"_disk2","bdf":"0000:00:04.0"}
```

liuw commented:
Changing this to WIP so CI runs are skipped.

liuw commented on vmm/src/vm_config.rs:
```diff
@@ -547,6 +547,12 @@ pub struct TpmConfig {
 }
 
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+pub struct LandlockConfig {
```
This also needs to be exposed via the http endpoint so that Kata and other orchestration layer can use it.

praveen-pk replied:
Agreed. I have to figure how I can order this correctly. Meaning, I need to accept LandlockConfigs sent before `vm_create` and ignore configs sent after vm_create.

praveen-pk replied:
On second thought, I don't see a lot of value in exposing this via http endpoint. Almost all the config specific rules are self-generated and handled. The only additional rules users have to pass are those for hot-adds. 

The Orchestration layer, can only pass landlock-rules before `vm_create`. It seems fair to let users start cloud-hypervisor with proper `--landlock-rules` and not worry about receiving additional rules via http endpoint.

liuw replied:
You can check with the Mariner team or the Kata community how this can be configured in their environment.

rbradford replied:
> This also needs to be exposed via the http endpoint so that Kata and other orchestration layer can use it.

I see this like seccomp - this is vmm configuration not vm configuration and only the latter is exposed via the HTTP or D-Bus API.

praveen-pk commented on vmm/src/lib.rs:
```diff
@@ -1197,12 +1236,72 @@ impl Vmm {
+// Add rules to Landlock ruleset based on VM Configuration and LandlockConfig
+fn apply_landlock(
+    landlock_config: Option<Vec<LandlockConfig>>,
+    vm_config: Option<Arc<Mutex<VmConfig>>>,
```
`apply_landlock` is called vm_create, which verifies VMConfig is not None.

As you suggested, there is not need to carry Option here. I will clean it up in next revision.

praveen-pk commented on vmm/src/lib.rs:
```diff
@@ -1197,12 +1236,72 @@ impl Vmm {
+    }
+
+    let payload_path = vm_config
+        .as_ref()
```
gotcha!

praveen-pk commented on vmm/src/lib.rs:
```diff
@@ -1197,12 +1236,72 @@ impl Vmm {
+        .payload
+        .as_ref()
+        .unwrap()
+        .kernel
```
Will check, if kernel exists here.

liuw replied:
Or rather you should check what kind of payload there is and make sure the the kernel + initrd or the igvm file is accessible.

praveen-pk replied:
checked now.

praveen-pk commented on vmm/src/lib.rs:
```diff
@@ -608,6 +635,14 @@ impl Vmm {
                                 }
                             }
+                            if landlock_enable{
+                                let _ = Landlock::new().unwrap().restrict_self().map_err(Error::ApplyLandlock).map_err(|e| {
```
Essentially an empty ruleset. After running this step, this thread will not be able to open any new files, as the ruleset is empty.

The thread will continue having access to all the FDs opened until this step.

praveen-pk commented on vmm/src/lib.rs:
```diff
@@ -608,6 +635,14 @@ impl Vmm {
+                                let _ = Landlock::new().unwrap().restrict_self().map_err(Error::ApplyLandlock).map_err(|e| {
+                                    error!("Error applying landlock to signal handler thread: {:?}", e);
+                                    exit_evt.write(1).ok();
+                                    return;
```
I followed the same logic as the seccomp filter failure above. I will double check this.

praveen-pk commented on src/main.rs:
```diff
@@ -255,6 +256,20 @@ fn create_app(default_vcpus: String, default_memory: String, default_rng: String
+            Arg::new("landlock")
+                .long("landlock")
+                .num_args(1)
+                .value_parser(["true", "false"])
```
During development, I configured this option to be `false` by default. Once landlock support is stable, we can enable this by default, if the host supports it.

liuw replied:
If we want to enable it by default, then all existing functionalities should continue to work. As I understand it, the code cannot handle hotplug yet?

praveen-pk replied:
Hotplug is supported, if the paths from the hot-plugged resources are passed using LandlockConfig, before starting the VM.

russell-islam commented:
@praveen-pk  Any plan to write documentation for this feature?

l0kod commented:
Instead of the `landlock-rules` argument which is orthogonal to CH's configuration, would it be possible to have a name dedicated to a directory of dynamic disk images instead? Using a Landlock-specific option could surprise and maybe make users reluctant to use it because they don't know the goal of this option (e.g. "Should I make it `rwx` or just `rw`?"), whereas they would know the meaning of an `image-directory` or `hotplug-directory` argument for instance. Such a clear semantic could also help for the documentation. Another advantage would be to not have a sub-argument/parsing dedicated to access rights (that will grow over time), which would also simplify the code. Because the semantic would be clear (i.e. read and write a disk image), we could also take advantage of future Landlock access rights in a best-effort way to protect users as much as possible (e.g. if the running kernel supports it, restrict `truncate`, `ioctl`...).

praveen-pk commented:
> advantage of future Landlock access

disk-images is just one of the arguments that takes file paths. There are 13 other arguments that take file paths. So, I kept this argument generic to allow users passing a dir/file path that will be used in any of the other arguments.

praveen-pk commented:
> @praveen-pk Any plan to write documentation for this feature?

Yes. Once the implementation is complete.

praveen-pk commented on src/main.rs:
```diff
@@ -255,6 +256,20 @@ fn create_app(default_vcpus: String, default_memory: String, default_rng: String
         )
+        .arg(
+            Arg::new("landlock")
+                .long("landlock")
```
updated

praveen-pk commented on src/main.rs:
```diff
@@ -603,6 +618,29 @@ fn start_vmm(cmd_arguments: ArgMatches) -> Result<Option<String>, Error> {
+
+    let mut landlock_config = None;
+    if let Some(landlock_rules) = cmd_arguments.get_many::<String>("landlock-rules") {
+        let mut landlock_config_from_args = Vec::new();
```
adopted Iterator.

praveen-pk commented:
With the latest version: 
* I moved LandlockConfig into VMConfig. This allows simpler code path during `vm_restore`.
* Enable landlock in `vm_restore` path. While testing this I encountered https://github.com/cloud-hypervisor/cloud-hypervisor/issues/6265.
* I am next looking at enabling landlock with PCI device passthrough.

praveen-pk commented:
With today's push:
* PCI passthrough is working as expected. As mentioned in https://github.com/cloud-hypervisor/cloud-hypervisor/issues/5170#issuecomment-1986497233, this need additional host side configuration.
* Now running integration tests to check if I missed anything.

liuw commented:
> With today's push:
> 
> * PCI passthrough is working as expected. As mentioned in [Sandboxing Cloud Hypervisor #5170 (comment)](https://github.com/cloud-hypervisor/cloud-hypervisor/issues/5170#issuecomment-1986497233), this need additional host side configuration.
> * Now running integration tests to check if I missed anything.

ARM64 tests are failing.

praveen-pk commented:
Can someone help re-trigger the failing ARM tests? On my local Raspberry pi 4, I see 
```
$ scripts/dev_cli.sh tests --unit --libc musl
20240229-0: Pulling from cloud-hypervisor/cloud-hypervisor
Digest: sha256:a459b28e7129e8c55113183b7f7ae8a075e6143df0e97165953b7267113e5c06
Status: Image is up to date for ghcr.io/cloud-hypervisor/cloud-hypervisor:20240229-0
ghcr.io/cloud-hypervisor/cloud-hypervisor:20240229-0
[Cloud Hypervisor] Running unit tests for aarch64-unknown-linux-musl...
   Compiling proc-macro2 v1.0.78
   Compiling unicode-ident v1.0.12
   Compiling libc v0.2.153
   Compiling serde v1.0.196
   Compiling bitflags v1.3.2
   Compiling thiserror v1.0.52
   Compiling bitflags v2.4.2
   Compiling log v0.4.20
   Compiling arc-swap v1.6.0
   Compiling cfg-if v1.0.0
   Compiling syn v1.0.109
   Compiling quote v1.0.35
   Compiling byteorder v1.4.3
   Compiling syn v2.0.48
   Compiling getrandom v0.2.11
   Compiling crc64 v2.0.0
   Compiling serde_json v1.0.109
   Compiling itoa v1.0.9
   Compiling ryu v1.0.15
   Compiling anyhow v1.0.79
   Compiling memchr v2.5.0
   Compiling ident_case v1.0.1
   Compiling fnv v1.0.7
   Compiling strsim v0.10.0
   Compiling kvm-ioctls v0.16.0
   Compiling virtio-bindings v0.2.2
   Compiling autocfg v1.1.0
   Compiling uuid v1.3.4
   Compiling lock_api v0.4.10
   Compiling epoll v4.3.3
   Compiling rustix v0.38.25
   Compiling scopeguard v1.2.0
   Compiling linux-raw-sys v0.4.13
   Compiling darling_core v0.20.6
   Compiling smallvec v1.13.1
   Compiling cc v1.0.83
   Compiling once_cell v1.19.0
   Compiling pkg-config v0.3.30
   Compiling utf8parse v0.2.1
   Compiling anstyle-parse v0.2.1
   Compiling vm-fdt v0.2.0 (https://github.com/rust-vmm/vm-fdt?branch=main#77212bd0)
   Compiling vcpkg v0.2.15
   Compiling serde_derive v1.0.196
   Compiling thiserror-impl v1.0.52
   Compiling versionize_derive v0.1.6 (https://github.com/cloud-hypervisor/versionize_derive?branch=ch-0.1.6#7906da99)
   Compiling darling_macro v0.20.6
   Compiling fdt v0.1.5
   Compiling colorchoice v1.0.0
   Compiling darling v0.20.6
   Compiling serde_with_macros v3.4.0
   Compiling anstyle-query v1.0.2
   Compiling anstyle v1.0.6
   Compiling anstream v0.6.13
   Compiling vm-memory v0.14.0
   Compiling terminal_size v0.3.0
   Compiling clap_lex v0.7.0
   Compiling io-uring v0.6.2
   Compiling linux-loader v0.11.0
   Compiling strsim v0.11.0
   Compiling pin-project-internal v1.1.4
   Compiling clap_builder v4.5.1
   Compiling debug-helper v0.3.13
   Compiling crc-any v2.4.4
   Compiling pin-project v1.1.4
   Compiling remain v0.2.11
   Compiling openssl-src v300.1.5+3.1.3
   Compiling spin v0.9.8
   Compiling nanorand v0.7.0
   Compiling clap v4.5.1
   Compiling futures-sink v0.3.30
   Compiling futures-core v0.3.30
   Compiling flume v0.10.14
   Compiling openssl-sys v0.9.99
   Compiling aho-corasick v1.0.2
   Compiling libz-sys v1.1.15
   Compiling option_parser v0.1.0 (/cloud-hypervisor/option_parser)
   Compiling regex-syntax v0.7.5
   Compiling vmm-sys-util v0.12.1
   Compiling bincode v1.3.3
   Compiling versionize v0.2.0
   Compiling vfio-bindings v0.4.0 (https://github.com/rust-vmm/vfio?branch=main#0daff4d4)
   Compiling kvm-bindings v0.7.0 (https://github.com/cloud-hypervisor/kvm-bindings?branch=ch-v0.7.0#2dcf85d4)
   Compiling virtio-queue v0.11.0
   Compiling serde_with v3.4.0
   Compiling vfio-ioctls v0.2.0 (https://github.com/rust-vmm/vfio?branch=main#0daff4d4)
   Compiling vm-virtio v0.1.0 (/cloud-hypervisor/vm-virtio)
   Compiling net_gen v0.1.0 (/cloud-hypervisor/net_gen)
   Compiling vm-migration v0.1.0 (/cloud-hypervisor/vm-migration)
   Compiling vhost v0.10.0
   Compiling hypervisor v0.1.0 (/cloud-hypervisor/hypervisor)
   Compiling rate_limiter v0.1.0 (/cloud-hypervisor/rate_limiter)
   Compiling vfio_user v0.1.0 (https://github.com/rust-vmm/vfio-user?branch=main#a1f6e528)
   Compiling net_util v0.1.0 (/cloud-hypervisor/net_util)
   Compiling arch v0.1.0 (/cloud-hypervisor/arch)
   Compiling vm-device v0.1.0 (/cloud-hypervisor/vm-device)
   Compiling block v0.1.0 (/cloud-hypervisor/block)
   Compiling vm-allocator v0.1.0 (/cloud-hypervisor/vm-allocator)
   Compiling regex-automata v0.3.7
   Compiling parking_lot_core v0.8.6
   Compiling pci v0.1.0 (/cloud-hypervisor/pci)
   Compiling regex v1.9.4
   Compiling event_monitor v0.1.0 (/cloud-hypervisor/event_monitor)
   Compiling zerocopy-derive v0.7.32
   Compiling libssh2-sys v0.3.0
   Compiling is-terminal v0.4.9
   Compiling instant v0.1.12
   Compiling humantime v2.1.0
   Compiling termcolor v1.2.0
   Compiling env_logger v0.10.1
   Compiling zerocopy v0.7.32
   Compiling no-std-net v0.6.0
   Compiling option-ext v0.2.0
   Compiling dirs-sys v0.4.1
   Compiling pnet_base v0.34.0
   Compiling acpi_tables v0.1.0 (https://github.com/rust-vmm/acpi_tables?branch=main#76e8552f)
   Compiling parking_lot v0.11.2
   Compiling vhost-user-backend v0.13.1
   Compiling tpm v0.1.0 (/cloud-hypervisor/tpm)
   Compiling enumflags2_derive v0.7.8
   Compiling seccompiler v0.4.0
   Compiling serial_buffer v0.1.0 (/cloud-hypervisor/serial_buffer)
   Compiling micro_http v0.1.0 (https://github.com/firecracker-microvm/micro-http?branch=main#e75dfa1e)
   Compiling signal-hook v0.3.17
   Compiling enumflags2 v0.7.8
   Compiling dirs v5.0.1
   Compiling signal-hook-registry v1.4.1
   Compiling wait-timeout v0.2.0
   Compiling landlock v0.3.0
   Compiling virtio-devices v0.1.0 (/cloud-hypervisor/virtio-devices)
   Compiling devices v0.1.0 (/cloud-hypervisor/devices)
   Compiling pnet_macros v0.34.0
   Compiling pnet_macros_support v0.34.0
   Compiling tracer v0.1.0 (/cloud-hypervisor/tracer)
   Compiling pnet_sys v0.34.0
   Compiling vhost_user_block v0.1.0 (/cloud-hypervisor/vhost_user_block)
   Compiling vhost_user_net v0.1.0 (/cloud-hypervisor/vhost_user_net)
   Compiling ipnetwork v0.20.0
   Compiling pnet v0.34.0
   Compiling cloud-hypervisor v38.0.0 (/cloud-hypervisor)
   Compiling pnet_packet v0.34.0
   Compiling pnet_datalink v0.34.0
   Compiling vmm v0.1.0 (/cloud-hypervisor/vmm)
   Compiling pnet_transport v0.34.0
   Compiling api_client v0.1.0 (/cloud-hypervisor/api_client)
   Compiling performance-metrics v0.1.0 (/cloud-hypervisor/performance-metrics)
    Compiling ssh2 v0.9.4
   Compiling test_infra v0.1.0 (/cloud-hypervisor/test_infra)
    Finished test [unoptimized + debuginfo] target(s) in 11m 10s
     Running unittests src/lib.rs (target/aarch64-unknown-linux-musl/debug/deps/api_client-0cb4984772b73665)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running unittests src/lib.rs (target/aarch64-unknown-linux-musl/debug/deps/arch-7181766c950348f5)

running 1 test
test aarch64::tests::test_arch_memory_regions_dram ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running unittests src/lib.rs (target/aarch64-unknown-linux-musl/debug/deps/block-23edb28dd45c3200)

running 36 tests
test qcow::tests::default_header_v3 ... ok
test qcow::tests::header_read ... ok
test qcow::tests::header_v2_with_backing ... ok
test qcow::tests::header_v3_with_backing ... ok
test qcow::tests::invalid_cluster_bits ... ok
test qcow::tests::invalid_magic ... ok
test qcow::tests::invalid_refcount_order ... ok
test qcow::tests::default_header_v2 ... ok
test qcow::tests::read_small_buffer ... ok
test qcow::tests::offset_write_read ... ok
test qcow::tests::rebuild_refcounts ... ok
test qcow::tests::seek_data ... ok
test qcow::tests::seek_data_direct ... ok
test qcow::tests::seek_hole ... ok
test qcow::tests::replay_ext4 ... ok
test qcow::tests::seek_hole_direct ... ok
test qcow::tests::test_header ... ok
test qcow::tests::test_header_1_tb_file_min_cluster ... ok
test qcow::tests::test_header_crazy_file_size_rejected ... ok
test qcow::tests::test_header_huge_file ... ok
test qcow::tests::test_header_huge_num_refcounts ... ok
test qcow::tests::test_header_huge_refcount_offset ... ok
test qcow::tests::test_huge_l1_table ... ok
test qcow::tests::write_read_start ... ok
test qcow::tests::test_header_1_tb_file ... ok
test qcow::tests::write_read_start_backing_overlap ... ok
test qcow::tests::combo_write_read ... ok
test qcow::tests::combo_write_read_direct ... ok
test qcow::tests::write_read_start_backing_v3 ... ok
test qcow::vec_cache::tests::evicts_when_full ... ok
test vhd::tests::test_check_vhd_footer ... ok
test vhd::tests::test_is_fixed_vhd ... ok
test vhd::tests::test_is_not_fixed_vhd ... ok
test qcow::tests::write_read_start_backing_v2 ... ok
test qcow::tests::write_zeroes_full_cluster ... ok
test qcow::tests::write_zeroes_read ... ok
```

gets past build and gets to running unit tests. Also the failure reported doesn't seem to be related the PR.

l0kod commented on vmm/src/api/http/mod.rs:
```diff
@@ -327,6 +329,18 @@ fn start_http_thread(
 
+            if landlock_enable {
+                Landlock::new()
+                    .unwrap()
```
Yes, all `unwrap()` should be replaced with proper error handling.

praveen-pk replied:
Thanks for catching this. I will fix this and push an update

praveen-pk replied:
done

l0kod commented on vmm/src/landlock.rs:
```diff
@@ -0,0 +1,130 @@
+pub const EXECUTE: u8 = 1 << 2;
+
+#[derive(Debug, Error)]
+pub enum LandlockError {
```
Do we really need this new error type? `RulesetError` should already contain the required information/types.

praveen-pk replied:
For the most part, the errors here map to ones in `RulesetError`. The only case that doesn't have an error in `RulesetError` is when the ruleset is empty. Such a case doesn't map to a proper error in RulesetError. 

If you can add such an error case, I will drop LandlockError here.

l0kod replied:
> The only case that doesn't have an error in `RulesetError` is when the ruleset is empty. Such a case doesn't map to a proper error in RulesetError.
> 
> If you can add such an error case, I will drop LandlockError here.

An empty ruleset is not an error because it can be used to just drop a set of access rights (without exception), for instance when only using inherited file descriptors.

l0kod commented on vmm/src/landlock.rs:
```diff
@@ -0,0 +1,130 @@
+}
+
+impl Landlock {
+    pub fn new() -> Result<Landlock, LandlockError> {
```
Shouldn't `new()` without argument be replaced with `Default::default()`?

praveen-pk replied:
From what I checked `default` is implemented only for `Ruleset`, which I use it in this method.

I am trying to store, a `RulesetCreated` object in `Landlock`. I don't follow how I can use `Default::default()` to replace this `new()` method.

l0kod replied:
I was just thinking about implementing the `Default` trait for your `Landlock` struct, and then use the `default()` call name instead of `new()`. See https://rust-lang.github.io/rust-clippy/stable/index.html#/new_without_default

praveen-pk replied:
@l0kod, I investigated a bit. From what I see `Default` is sensible when there are no failure cases. In this case, Landlock invokes `ruleset.create()` which could fail, if the underlying kernel does not support it. 

If I implement `Default` to Landlock, it would look like below:
```
impl Default for Landlock {
    fn default() -> Self {
        Landlock::new().unwrap()
    }
}
```
which relies on `unwrap()`, which isn't good. Please let me know if I missed anything.

l0kod replied:
Correct, `new()` is fine.

l0kod commented on vmm/src/landlock.rs:
```diff
@@ -0,0 +1,130 @@
+            self.ruleset
+                .as_mut()
+                .unwrap()
+                .add_rule(rule.unwrap())
```
Why not using `add_rules()`?

praveen-pk replied:
I am adding a single rule in this method and so not using `add_rules()`. I use the indirection via `path_beneath_rules` only to handle file vs dir permissions. In order to use `path_beneath_rules` I create a `vec` of single path and pass it.

praveen-pk replied:
I re-checked this and add_rules seems cleaner:

```
let paths = vec![path.clone()];
        let path_beneath_rules = path_beneath_rules(paths, access);
        if let Some(ruleset) = &mut self.ruleset {
            ruleset.add_rules(path_beneath_rules)?;
        }
```
I will push this in next update.

praveen-pk replied:
done

l0kod commented on vmm/src/landlock.rs:
```diff
@@ -0,0 +1,130 @@
+        let path_beneath_rules = path_beneath_rules(paths, access);
+        for rule in path_beneath_rules {
+            self.ruleset
+                .as_mut()
```
Do we really need an explicit call to `as_mut()`?

praveen-pk replied:
As ruleset is of type `Option<RulesetCreated>`, I need to use `as_mut()`. Without it I get the following error:
```
error[E0507]: cannot move out of `self.ruleset` which is behind a mutable reference
   --> vmm/src/landlock.rs:77:13
    |
77  |             self.ruleset
    |             ^^^^^^^^^^^^
    |             |
    |             help: consider calling `.as_ref()` or `.as_mut()` to borrow the type's contents
    |             move occurs because `self.ruleset` has type `std::option::Option<RulesetCreated>`, which does not implement the `Copy` trait
78  |                 .unwrap()
    |                  -------- `self.ruleset` moved due to this method call
    |
note: `std::option::Option::<T>::unwrap` takes ownership of the receiver `self`, which moves `self.ruleset`
```

praveen-pk replied:
As shown in previous comment, with the use of add_rules, don't need `as_mut()`.

praveen-pk replied:
done

l0kod commented on vmm/src/landlock.rs:
```diff
@@ -0,0 +1,130 @@
+        Ok(())
+    }
+
+    fn flags_to_access(&self, flags: u8) -> BitFlags<AccessFs> {
```
Why using `u8` instead of an `enum` or a dedicated `struct` (following the builder pattern with a `read()` and `write()` method to change its internal private state)?

praveen-pk replied:
`flags` here is an `OR` of of `READ`, `WRITE` and `EXECUTE`, which are declared as bits. 
After I read cmdline args like `--landlock-rules path=<a/b/c>,flags=rwx`, I translate the flags into transcode `rwx` to bitwise ORs of READ, WRITE, EXECUTE and pass it here.

rbradford replied:
https://crates.io/crates/bitflags is a good shout for this

l0kod replied:
Yes, but from an API point of view, the issue with this design is that unknown bits are ignored (at run time) instead of being denied by design (at build time).

praveen-pk replied:
If unknown bits are passed at cmdline, cloud-hypervisor errors out in `LandlockConfig` -> `parse`.

praveen-pk replied:
updated to use bitflags.

l0kod reviewed (commented):
I really like the way Landlock is applied with all the `*Config::landlock_apply()`

l0kod commented on vmm/src/config.rs:
```diff
@@ -2100,6 +2128,51 @@ impl TpmConfig {
+        let mut path_flags: u8 = 0;
+        if let Some(flags) = parser.get("flags") {
+            for c in flags.chars() {
+                match c {
```
I'd suggest to be conservative and only allow one character of each to not accept cases like "rwxxxxrr".

praveen-pk replied:
I will address this

praveen-pk replied:
done

l0kod commented on vmm/src/landlock.rs:
```diff
@@ -0,0 +1,124 @@
+        // have to worry about the type of the path.
+        let paths = vec![path.clone()];
+        let path_beneath_rules = path_beneath_rules(paths, access);
+        if let Some(ruleset) = &mut self.ruleset {
```
Oh, I now understand your use of `RulesetConsumed`. Why do you need `self.ruleset` to be an `Option`? The library was designed to avoid as much as possible run time checks such as this one. This is why some methods consume the `self` object instead of passing it as a reference, and also why there is the `Compatible` trait which should handle all cases when Landlock is not supported.

praveen-pk replied:
I made `self.ruleset` an `Option` because the ruleset is consumed whenever `add_rule` is called. I see `add_rule` returns the ruleset after consuming it. I can drop `Option` for self.ruleset and update it after every `add_rule`.

praveen-pk replied:
After investigating this a bit, I realized `add_rules` consumes the passed `ruleset` and returns a new one. I updated the code to handle this properly.

This in turn requires me to consume `landlock` object and return a new one from the methods defined in my landlock.rs. This look a bit odd. Any suggestions on how to better handle this?

l0kod replied:
Using a mutable reference should work because [`RulesetCreatedAttr` is implemented for `&mut RulesetCreated` too](https://landlock.io/rust-landlock/landlock/trait.RulesetCreatedAttr.html#impl-RulesetCreatedAttr-for-%26mut+RulesetCreated).

praveen-pk replied:
updated

praveen-pk commented on vmm/src/vm_config.rs:
```diff
@@ -440,6 +515,21 @@ pub struct DeviceConfig {
+        let device_flags = landlock::Perms::READ | landlock::Perms::WRITE;
+        let mut landlock = landlock;
+        //TODO: Fix this unwrap. It should be replaced with proper RulesetError
+        let device_path = fs::read_link(self.path).unwrap();
```
@l0kod I dropped LandlockError from the previous version. The only place this is causing an issue is here.

As this check returns an `io::Error`, I cannot return a `RulesetError` from this method. I can think of below possibilities to address this:

1) re-introduce LandlockError to properly address this case
2) Extend RulesetError with another error which takes io::Error
3) forcefully shoehorn this error into one of the existing RulesetError. I checked and couldn't find anything sensible.

l0kod replied:
Looks good. We may want to update rust-landlock but for now let's stick to this.

liuw commented on vmm/src/vm_config.rs:
```diff
@@ -115,7 +120,16 @@ pub struct MemoryZoneConfig {
 }
-
+impl MemoryZoneConfig {
+    pub fn apply_landlock(self, landlock: Landlock) -> LandlockResult<Landlock> {
```
Why does this function (and other similar ones) need to take a `self`? If you take a reference to then you can skip all the `clone` calls below?

Also, since this is the same function that gets called all the time, you can define a trait as part of the LandLock infrastructure in CH. And then you may do dynamic dispatch without writing code whenever a new configuration option is added.

liuw replied:
I thought about the second point. It is not really hassle free. Even with a default implementation you still need to tag all the configuration structures. That's a bit tedious.

praveen-pk replied:
> Why does this function (and other similar ones) need to take a self? If you take a reference to then you can skip all the clone calls below?

This is what I had in my previous version. I pushed this modification in response to comments from Mickael. I agree this is not clean. I will sync up with Mickael on this change before I push my next version.

praveen-pk commented:
With the latest version:
* Addressed all the pending comments
* Introduced a fix to support Landlock while using pty devices
* Fixed a couple of integration tests to run with landlock enabled

Next:
* I still have a few integration tests failing with Landlock enabled. I will fix the rest in my next update.

liuw commented on vmm/src/lib.rs:
```diff
@@ -1202,12 +1243,58 @@ impl Vmm {
 }
 
+/* Create ptys here and add the slave paths to Serial, Console and DebugConsole
+    configs. This allows apply_landlok to add the slave paths to landlock rules
```
Wrong comment style.

liuw commented on vmm/src/lib.rs:
```diff
@@ -1202,12 +1243,58 @@ impl Vmm {
+    later.
+*/
+fn create_ptys(config: Arc<Mutex<VmConfig>>) -> result::Result<(), LandlockError> {
+    if config.lock().unwrap().console.mode == ConsoleOutputMode::Pty {
```
You can do `let config = config.lock().unwrap()` once at the start of the function.

liuw commented on vmm/src/vm_config.rs:
```diff
@@ -115,7 +121,15 @@ pub struct MemoryZoneConfig {
 }
-
+impl MemoryZoneConfig {
+    pub fn apply_landlock(self, landlock: &mut Landlock) -> LandlockResult<()> {
```
The first parameter should be `&self`. There is no need to consume the structure.

liuw commented on vmm/src/vm_config.rs:
```diff
@@ -628,4 +796,96 @@ pub struct VmConfig {
+
+        if let Some(mem_zones) = self.memory.zones.as_ref() {
+            for zone in mem_zones.iter() {
+                zone.clone().apply_landlock(&mut landlock)?;
```
Please drop all the `clone`s. See my other comment about `&self`.

praveen-pk commented:
#6403 is required to pre-create console devices before `vm_create`. This is required to enable Landlock while using console devices.

praveen-pk commented:
```$ cargo test --package rate_limiter --lib -- tests::test_rate_limiter_bandwidth --exact --show-output
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.09s
     Running unittests src/lib.rs (target/debug/deps/rate_limiter-32a61e085d769eaa)
                                                                                                                                                running 1 test
test tests::test_rate_limiter_bandwidth ... ok

successes:

successes:
    tests::test_rate_limiter_bandwidth
                                                                                                                                                test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 17 filtered out; finished in 0.10s
```
test_rate_limiter_bandwidth is passing locally. This could be pipeline noise.

praveen-pk commented:
@cloud-hypervisor/cloud-hypervisor-maintainers how can I customize the configuration of VMs running the cloud-hypervisor integration pipelines. I need to enable Landlock on these VMs to get landlock integration tests to pass.

rbradford commented:
> @cloud-hypervisor/cloud-hypervisor-maintainers how can I customize the configuration of VMs running the cloud-hypervisor integration pipelines. I need to enable Landlock on these VMs to get landlock integration tests to pass.

Please can you define "enable landlock" ? Is that a custom kernel or?

praveen-pk commented:
> 
> Please can you define "enable landlock" ? Is that a custom kernel or?

I need to enable `CONFIG_SECURITY_LANDLOCK` in the kernel and pass `lsm=landlock` via kernel cmdline during boot.

rbradford commented:
> > Please can you define "enable landlock" ? Is that a custom kernel or?
> 
> I need to enable `CONFIG_SECURITY_LANDLOCK` in the kernel and pass `lsm=landlock` via kernel cmdline during boot.

I'm not sure that's going to be doable with on on demand VMs as that would mean maintaining a custom disk image on Azure which creates a non insubstantial maintenance burden. Several alternatives come to mind:

a.) We test it nested inside another CH VM running the custom CH kernel - where that option and boot time option can be easily enabled - we already do nesting like that with `test_vfio`
b.) We run those tests only on either an existing dedicated worker or, better, a new dedicated worker - this would mean testing would only happen on the MQ.
c.) We don't test that functionality (this wouldn't be unique and the MSHV and TDX functionality is not tested by the CI.)

praveen-pk commented:
> a.) We test it nested inside another CH VM running the custom CH kernel - where that option and boot time option can be easily enabled - we already do nesting like that with `test_vfio` b.) We run those tests only on either an existing dedicated worker or, better, a new dedicated worker - this would mean testing would only happen on the MQ. c.) We don't test that functionality (this wouldn't be unique and the MSHV and TDX functionality is not tested by the CI.)


Could you please explain what MQ is? I see MQ being used as an abbreviation for multi-queue in code, but I don't follow the relation to CI here. By dedicated worker, do you mean a custom VM specifically for landlock test?

rbradford commented:
> > a.) We test it nested inside another CH VM running the custom CH kernel - where that option and boot time option can be easily enabled - we already do nesting like that with `test_vfio` b.) We run those tests only on either an existing dedicated worker or, better, a new dedicated worker - this would mean testing would only happen on the MQ. c.) We don't test that functionality (this wouldn't be unique and the MSHV and TDX functionality is not tested by the CI.)
> 
> Could you please explain what MQ is? I see MQ being used as an abbreviation for multi-queue in code, but I don't follow the relation to CI here. By dedicated worker, do you mean a custom VM specifically for landlock test?

MQ in this context is Merge Queue - we running many more tests only once the PR is added to the merge queue. Our CI is powered by self hosted GitHub actions workers that are either spawned on demand on Azure or via some fixed bare metal machines.

praveen-pk commented:
I drop the integration tests for this PR for now. At a later time I will re-write the tests using nested VM options.

rbradford commented:
> I'll drop the integration tests for this PR for now. At a later time I will re-write the tests using nested VM options.

That seems reasonable - thanks for your understanding!

l0kod commented:
> > > Please can you define "enable landlock" ? Is that a custom kernel or?
> > 
> > 
> > I need to enable `CONFIG_SECURITY_LANDLOCK` in the kernel and pass `lsm=landlock` via kernel cmdline during boot.
> 
> I'm not sure that's going to be doable with on on demand VMs as that would mean maintaining a custom disk image on Azure which creates a non insubstantial maintenance burden. Several alternatives come to mind:

Which kernel and/or distro is used? Landlock should already be enabled by default in most distros.

liuw commented:
> > > > Please can you define "enable landlock" ? Is that a custom kernel or?
> > > 
> > > 
> > > I need to enable `CONFIG_SECURITY_LANDLOCK` in the kernel and pass `lsm=landlock` via kernel cmdline during boot.
> > 
> > 
> > I'm not sure that's going to be doable with on on demand VMs as that would mean maintaining a custom disk image on Azure which creates a non insubstantial maintenance burden. Several alternatives come to mind:
> 
> Which kernel and/or distro is used? Landlock should already be enabled by default in most distros.



> > > > Please can you define "enable landlock" ? Is that a custom kernel or?
> > > 
> > > 
> > > I need to enable `CONFIG_SECURITY_LANDLOCK` in the kernel and pass `lsm=landlock` via kernel cmdline during boot.
> > 
> > 
> > I'm not sure that's going to be doable with on on demand VMs as that would mean maintaining a custom disk image on Azure which creates a non insubstantial maintenance burden. Several alternatives come to mind:
> 
> Which kernel and/or distro is used? Landlock should already be enabled by default in most distros.

Ubuntu Jammy per the CI configuration.

l0kod commented:
> Ubuntu Jammy per the CI configuration.

Landlock is enabled by default on this distro, so no need for a specific kernel.

liuw commented:
> > Ubuntu Jammy per the CI configuration.
> 
> Landlock is enabled by default on this distro, so no need for a specific kernel.

My reading of Praveen's comment is that he also needs to add `lsm=landlock` to the kernel command line. That doesn't seem possible with a GitHub runner.

praveen-pk commented:
> > Ubuntu Jammy per the CI configuration.
> 
> Landlock is enabled by default on this distro, so no need for a specific kernel.

Thanks for catching this. I confirmed that Jammy cloud VMs do have Landlock properly enabled. 

The last time the CI pipeline failed for this PR, Landlock test failed for `aarch64`, which seems to be using Focal Images. If x86_64 tests use Jammy, I can enable Landlock tests for x86_64 only for now.

praveen-pk commented:
I re-enabled Landlock tests only for x86_64 arch and they are all passing in CI pipeline:

```
test common_parallel::test_landlock ... ok
test common_parallel::test_disk_hotplug_with_landlock ... ok


test live_migration::live_migration_parallel::test_live_migration_with_landlock ... ok

```

liuw reviewed (commented):
Mostly looks good to me, just a few minor issues and a question I have.

rbradford commented:
> I re-enabled Landlock tests only for x86_64 arch and they are all passing in CI pipeline:
> 
> ```
> test common_parallel::test_landlock ... ok
> test common_parallel::test_disk_hotplug_with_landlock ... ok
> 
> 
> test live_migration::live_migration_parallel::test_live_migration_with_landlock ... ok
> ```

FWIW - I upgraded the ARM64 machine to Jammy now.

l0kod commented:
> > > Ubuntu Jammy per the CI configuration.
> > 
> > 
> > Landlock is enabled by default on this distro, so no need for a specific kernel.
> 
> My reading of Praveen's comment is that he also needs to add `lsm=landlock` to the kernel command line. That doesn't seem possible with a GitHub runner.

This is only needed when Landlock is built into the kernel but not enabled by default. There is no good reason for this to happen.

praveen-pk commented on docs/landlock.md:
```diff
@@ -0,0 +1,76 @@
+	--cpus boot=4 \
+	--memory size=1024M \
+	--net "tap=,mac=,ip=,mask=" \
+    --landlock \
```
fixed

praveen-pk commented on tests/integration.rs:
```diff
@@ -5190,6 +5195,70 @@ mod common_parallel {
+                Some(format!("path={},id=test0", blk_file_path.to_str().unwrap()).as_str()),
+            );
+            // This will fail, as the path for the hotplug disk is not pre-added
+            assert!(!cmd_fail);
```
I hear you. 
I dropped the variable to reduce any confusion, following other similar uses in this module.

praveen-pk commented on src/main.rs:
```diff
@@ -634,6 +655,8 @@ fn start_vmm(cmd_arguments: ArgMatches) -> Result<Option<String>, Error> {
     let vm_debug_evt = EventFd::new(EFD_NONBLOCK).map_err(Error::CreateDebugEventFd)?;
 
     let exit_evt = EventFd::new(EFD_NONBLOCK).map_err(Error::CreateExitEventFd)?;
+    // TODO: landlock_enable is being parsed in 2 places: here and VMparams:parse() how to merge both?
```
I rechecked and this comment is not entirely true. I am only reading from parsed results(ArgMatches), instead of re-parsing the cmdline options. 

I see similar uses in other places, so I dropped the comment.

rbradford commented:
> sudo dmesg | grep -i landlock

So although I did upgrade the OS it looks like the ARM64 machine is running a custom 5.16 kernel. Would need to investigate that. For now it's fine to just run this on the x86-64 workers only.

praveen-pk commented:
> So although I did upgrade the OS it looks like the ARM64 machine is running a custom 5.16 kernel. Would need to investigate that. For now it's fine to just run this on the x86-64 workers only.

Gotcha! I enabled landlock tests for x86_64 only for now. Thanks!

l0kod commented on vmm/src/config.rs:
```diff
@@ -2301,6 +2329,59 @@ impl TpmConfig {
+
+        let mut path_perms = landlock::Perms::empty();
+        let flags = parser
+            .get("flags")
```
I suggest using "access" instead of "flags" (here and in the doc), which is self-explanatory.

l0kod commented on vmm/src/landlock.rs:
```diff
@@ -0,0 +1,113 @@
+        if perms & Perms::READ != Perms::empty() {
+            access |= AccessFs::from_read(ABI);
+        }
+        if perms & Perms::READ != Perms::empty() {
```
This should be `Perms::WRITE`. Is it possible to add a test for this kind of `Perms` mapping?

Right now, all modules allowing `Perms::READ` permissions are also allowing write access, so we should check that the current policy is correct.

Maybe the `Perms` type could just be replaced with `BitFlags<AccessFs>` and a `TryFrom<&str>` implementation to manage the "rw" string mapping? This should simplify the code. We can still use a global `ACCESS_FS_READ = AccessFs::from_read(ABI);` (instead of `Perms::READ`) variable to simplify module configuration though.

liuw reviewed (approved):
There is one open question whether we should expose landlock via the restful API. That can be resolved separately.

liuw commented:
@rbradford @likebreath @l0kod do you have more comments? If you're happy with the changes can you please give your approvals?

l0kod commented on vmm/src/landlock.rs:
```diff
@@ -0,0 +1,147 @@
+            match c {
+                'r' => access |= AccessFs::from_read(ABI),
+                'w' => access |= AccessFs::from_write(ABI),
+                'x' => access |= AccessFs::from_write(ABI),
```
`'x' => access |= AccessFs::Execute,`

praveen-pk replied:
I handled 'x' rather loosely here. I re-checked and I didn't see any places where cloud-hypervisor process executes a file. Cloud-hypervisor program itself is executed which happens at the start, but cloud-hypervisor process doesn't open another program to execute it. 

I will drop 'x' and simplify this code.

l0kod commented on vmm/src/landlock.rs:
```diff
@@ -0,0 +1,147 @@
+        let mut access = BitFlags::<AccessFs>::empty();
+        for c in s.chars() {
+            match c {
+                'r' => access |= AccessFs::from_read(ABI),
```
Because there is an explicit `x`, we should have `'r' => access |= AccessFs::from_read(ABI) & !AccessFs::Execute,`

praveen-pk replied:
dropped `x`

l0kod commented on vmm/src/landlock.rs:
```diff
@@ -0,0 +1,147 @@
+    fn try_from(s: &str) -> Result<LandlockAccess, LandlockError> {
+        let mut access = BitFlags::<AccessFs>::empty();
+        for c in s.chars() {
+            match c {
```
Please note that using `from_*(abi)` means that future version of Cloud Hypervisor may be required to stick to the same Landlock ABI to avoid unexpected denied access. I suggest to document that 'r', 'w', and 'x' will translate to more restrictions in the future (e.g. I think we'll want 'w' to deny xattr changes, which is not implemented yet), and users need to be aware of this semantic. For instance, it would make sense to document that these access right will be similar to unix permissions eventually.

praveen-pk replied:
I will add a blurb to this effect in landlock.md

praveen-pk replied:
Below is the blurb added to landlock.md:

```
Under the scope of read and write access Landlock could allow some additional accesses (eg: write access allows xattr changes). Eventually, Landlock will only allow accesses similar to unix permissions.
```
Please check if this needs to be updated.

l0kod commented on vmm/src/landlock.rs:
```diff
@@ -0,0 +1,147 @@
+                    ))
+                }
+            };
+        }
```
We should require `access` to not be empty.

praveen-pk replied:
https://github.com/praveen-pk/cloud-hypervisor/blob/e91a3836daa91132e6564d23d9ac0b557279cb3d/vmm/src/config.rs#L2353

This is checked while validating the inputs of `--landlock-rules` option.

praveen-pk replied:
`add_rule_with_access` is invoked in response to arguments passed via `--landlock-rules`. Just to be extra cautious, I added a case to error out if `access` is empty. I will push this in my next update.

l0kod commented on vmm/src/landlock.rs:
```diff
@@ -0,0 +1,147 @@
+    assert!(landlock_access.access == AccessFs::from_write(ABI) | AccessFs::from_read(ABI));
+
+    let landlock_access = LandlockAccess::try_from("").unwrap();
+    assert!(landlock_access.access == BitFlags::<AccessFs>::empty());
```
This should be an error.

praveen-pk replied:
I will update this check

praveen-pk replied:
fixed

l0kod commented on vmm/src/landlock.rs:
```diff
@@ -0,0 +1,147 @@
+#[test]
+fn test_try_from_perms() {
+    let landlock_access = LandlockAccess::try_from("rwx").unwrap();
+    assert!(landlock_access.access == AccessFs::from_all(ABI));
```
I suggest to have a list of explicit access rights instead of using `from_all()` in this test. This would require to update this test with a new Landlock ABI but the consequence will be visible with the code review.

praveen-pk replied:
The latest update doesn't use `from_all` anymore.

l0kod commented on vmm/src/landlock.rs:
```diff
@@ -0,0 +1,147 @@
+    pub fn add_rule_with_access(
+        &mut self,
+        path: PathBuf,
+        access: String,
```
You could use `access: S` where `S: AsRef<str>` to avoid all `.to_string()` calls.

praveen-pk replied:
as this is a single invocation, I will leave this as is.

l0kod replied:
> as this is a single invocation, I will leave this as is.

`add_rule_with_access()` is called 14 times: each `ApplyLandlock` implementations.

l0kod commented on vmm/src/landlock.rs:
```diff
@@ -0,0 +1,147 @@
+
+    pub fn add_rule_with_access(
+        &mut self,
+        path: PathBuf,
```
You could use `path: P`, where `P: Into<PathBuf>` to avoid all `.into()` calls.

praveen-pk replied:
I dropped `into` here, as it was redundant.

rbradford commented on vmm/src/lib.rs:
```diff
@@ -347,6 +355,17 @@ pub fn start_event_monitor_thread(
+                Landlock::new()
+                    .map_err(Error::ApplyLandlock)?
+                    .restrict_self()
+                    .map_err(Error::ApplyLandlock)
```
I think this should be two different errors to help trace them.

praveen-pk replied:
fixed

rbradford commented on vmm/src/landlock.rs:
```diff
@@ -0,0 +1,131 @@
+
+use std::convert::TryFrom;
+use thiserror::Error;
+
```
Don't use whitespace in imports - because cargo fmt can't then enforce ordering correctly.

praveen-pk replied:
fixed

rbradford reviewed (commented):
Thanks for working on this!

praveen-pk commented on vmm/src/config.rs:
```diff
@@ -537,6 +563,11 @@ impl<'a> VmParams<'a> {
         #[cfg(feature = "sev_snp")]
         let host_data = args.get_one::<String>("host-data").map(|x| x as &str);
+        let landlock_enable = args.get_flag("landlock");
+        let landlock_config: Option<Vec<&str>> = args
```
`--landlock-rules` accepts multiple rules like below.

```
--landlock-rules path="/path/to/file",access="rw" path="/path/to/dir",access="rw"
```

So, I made it plural. I will update the documentation with a similar example.

praveen-pk commented on docs/landlock.md:
```diff
@@ -0,0 +1,80 @@
+# Sandboxing using Landlock
+Landlock is a lightweight mechanism to allow unprivileged applications to
```
done

praveen-pk commented on docs/landlock.md:
```diff
@@ -0,0 +1,80 @@
+	--net "tap=,mac=,ip=,mask=" \
+	--landlock \
+	--landlock-rules path="/path/to/hotplug",access="rw"
+
```
Added an example with multiple rules.

praveen-pk commented on src/main.rs:
```diff
@@ -271,6 +274,24 @@ fn create_app(default_vcpus: String, default_memory: String, default_rng: String
                 .num_args(1..)
                 .group("vm-config"),
         )
+        .arg(
```
Your update captures my intent. Hope that works for you.

rbradford replied:
Thank you!

l0kod commented on vmm/src/landlock.rs:
```diff
@@ -0,0 +1,130 @@
+#[test]
+fn test_try_from_perms() {
+    let landlock_access = LandlockAccess::try_from("rw").unwrap();
+    assert!(landlock_access.access == AccessFs::from_read(ABI) | AccessFs::from_write(ABI));
```
My comment was to use explicit access rights instead of the `from_*()` helpers to help future reviews when a `from_read(v9)` will return something different than `from_read(v3)`.

praveen-pk replied:
Gotcha!

I like the idea of highlighting access right changes to reviewers with landlock version bumps. Addressed this in my latest update.

l0kod reviewed (commented):
A few cosmetic comments, but it looks good overall!

l0kod commented on vmm/src/landlock.rs:
```diff
@@ -0,0 +1,130 @@
+    pub(crate) fn add_rule_with_access(
+        &mut self,
+        path: PathBuf,
+        access: String,
```
Following https://github.com/cloud-hypervisor/cloud-hypervisor/pull/6214/files#diff-af381b88eea7e5e5d9bb6626898fab0a2127ae012f87424ea44b20049f9a5319

> > You could use `access: S` where `S: AsRef<str>` to avoid all `.to_string()` calls.


> as this is a single invocation, I will leave this as is.

`add_rule_with_access()` is called 14 times: each `ApplyLandlock` implementations.

praveen-pk replied:
Seems like I mis-understood your comment. I re-checked this code and realized I didn't need `to_string()` in the first place. Fixed in the latest update. Thanks for pointing it out.

liuw commented:
Github CI was stuck. Closing and reopening this PR triggers the CI again.

l0kod commented on vmm/src/config.rs:
```diff
@@ -2301,6 +2334,45 @@ impl TpmConfig {
 
+impl LandlockConfig {
+    pub const SYNTAX: &'static str = "Landlock parameters \
+        \"path=<path/to/{file/dir}>,access={r,w}\"";
```
This looks more correct: `access=[rw]`

praveen-pk replied:
The existing looks reasonable to me.
@cloud-hypervisor/cloud-hypervisor-maintainers  any thoughts on this?

liuw replied:
It should be '[rw]'.

rbradford commented on src/main.rs:
```diff
@@ -271,6 +276,24 @@ fn create_app(default_vcpus: String, default_memory: String, default_rng: String
+            .long("landlock-rules")
+            .help(config::LandlockConfig::SYNTAX)
+            .num_args(1..)
+            .group("vm-config"),
```
I think this should be "vmm-config" group ?

rbradford commented on vmm/Cargo.toml:
```diff
@@ -39,6 +39,7 @@ hex = { version = "0.4.3", optional = true }
 hypervisor = { path = "../hypervisor" }
 igvm = { version = "0.3.1", optional = true }
 igvm_defs = { version = "0.3.1", optional = true }
+landlock = "0.4"
```
This should be a precise 3 digit version - dependabot bumps versions for us.

praveen-pk replied:
fixed

rbradford commented on vmm/src/vm_config.rs:
```diff
@@ -645,4 +796,94 @@ pub struct VmConfig {
     pub preserved_fds: Option<Vec<i32>>,
+    #[serde(default)]
+    pub landlock_enable: bool,
+    pub landlock_config: Option<Vec<LandlockConfig>>,
```
I'm a bit confused if this config is here (in `VmConfig`) then it it VM configuration not VMM configuration (like seccomp) and then will be automatically exposed via the HTTP API. Can you clarify whether this is VM configuration or VMM configuration?

praveen-pk replied:
I initially started with `vmm-config` like seccomp, but later moved it to `VmConfig`. I moved to VmConfig to allow for `LandlockConfig` to serialized and de-seriliazed along with rest of the VM configs during live migration.

liuw replied:
It appears this is more than changing the group. If we don't want those to be part of VmConfig, then the application of the rules needs to be rewritten.

liuw commented on src/main.rs:
```diff
@@ -85,6 +86,10 @@ enum Error {
     LoggerSetup(log::SetLoggerError),
     #[error("Failed to gracefully shutdown http api: {0}")]
     HttpApiShutdown(#[source] vmm::Error),
+    #[error("Failed to create Landlock object")]
```
Both this and the next error should include the source.

praveen-pk replied:
fixed now.

liuw commented:
Praveen is away for a few weeks. I can address all the comments and push a new version if I have some free cycles.

praveen-pk commented:
> Praveen is away for a few weeks. I can address all the comments and push a new version if I have some free cycles.

Thanks Wei

praveen-pk commented on docs/landlock.md:
```diff
@@ -0,0 +1,85 @@
+compromised.
+
+Under the scope of `read` and `write` access, Landlock currently allows some
+additional accesses (eg: `write` access allows `xattr` changes). Eventually,
```
fixed

liuw commented:
Now the tip commit https://github.com/cloud-hypervisor/cloud-hypervisor/pull/6214/commits/f8a439a0b1553df151df30f19496d0f5203a2ecb hoists landlock options from `VmConfig` to VMM level.

If people don't like that it can be reverted or dropped easily.

rbradford reviewed (approved):
Looks fine to me - I don't want to create extra work by asking you to go back and apply the tip change into the corresponding commits.

l0kod reviewed (approved):
Excellent work @praveen-pk!
